src/pj_fwd.c

```diff
@@ -54,18 +54,46 @@ PJ_COORD pj_fwd_prepare (PJ *P, PJ_COORD coo) {
         if (lp.phi < -M_HALFPI)
             lp.phi = -M_HALFPI;
 
+
+        coo.lp = lp;
+
         /* If input latitude is geocentrical, convert to geographical */
         if (P->geoc)
             coo = proj_geoc_lat (P, PJ_INV, coo);
 
         /* Distance from central meridian, taking system zero meridian into account */
-        lp.lam = (lp.lam - P->from_greenwich) - P->lam0;
+        coo.lp.lam = (coo.lp.lam - P->from_greenwich) - P->lam0;
 
         /* Ensure longitude is in the -pi:pi range */
         if (0==P->over)
-            lp.lam = adjlon(lp.lam);
+            coo.lp.lam = adjlon(coo.lp.lam);
+
+        if (P->hgridshift)
+            coo = proj_trans (P->hgridshift, -1, coo);
+        else if (P->helmert) {
+            coo = proj_trans (P->cart,    1, coo);
+            coo = proj_trans (P->helmert, 1, coo);
+            coo = proj_trans (P->cart,   -1, coo);
+        }
+        if (P->vgridshift)
+            coo = proj_trans (P->vgridshift, 1, coo);
+        return coo;
+    }
 
-        coo.lp = lp;
+
+    if (P->left==PJ_IO_UNITS_CARTESIAN) {
+        /* Do we really support gridshifts on cartesian input? */
```

Comment:
> Do we really support gridshifts on cartesian input?

Not in "classic PROJ" mode. This can only occur if when using the deformation operation which is not compatible with classic PROJ.

Reply:
So this will first really be an issue to decide on when we introduce global `+inv` handling and/or a `proj=cartesian` (or `proj=geocent` in deprecated PROJ terms).

I'll leave the code in, but we will have to decide what we mean with this. I can imagine situations where it makes sense, but mostly for convenience (but then again, there is nothing in this PR which cannot be handled by properly constructed pipelines - it is by and large a matter of improving backward compatibility by providing some convenience functionality, for interoperability between new code and old parameter file systems)

Reply:
> (but then again, there is nothing in this PR which cannot be handled by properly constructed pipelines - it is by and large a matter of improving backward compatibility by providing some convenience functionality, for interoperability between new code and old parameter file systems)

That's my point. Keep it as simple as possible and don't open the door too much for new ideas based on the classic PROJ.4 idioms.

Reply:
OK, I get your point now, and I fully agree, so I will remove that chunk of code